### DISCUSSION TITLE
How to use fzf remotely through SSH, within command substitution?

### DISCUSSION THREAD

**Discussion by FranklinYu** [junegunn/fzf#3072] [Q&A] [answered]
This works as expected:

```sh
ssh -t example.com "echo -e 'a\nb' | fzf"
```

But I cannot capture the output with Command Substitution:

```sh
selected_value=$(ssh -t example.com "echo -e 'a\nb' | fzf")
```

Any idea why? And what’s the right usage?

**Comment by junegunn** [comment, marked as answer]
Do you have to run fzf remotely? If that is not the case, a simple workaround would be to run it locally.

```sh
selected_value=$(ssh example.com "echo -e 'a\nb'" | fzf)

```

**Reply by FranklinYu** [reply]
If I run it locally, does it mean that every preview command will be another SSH command? There are certainly workarounds (such as SSH multiplexing), but I’m curious why it doesn’t “Just Work”.

**Reply by junegunn** [reply]
Because ssh doesn't forward stderr and fzf renders its interface using stderr.

```sh
ssh -t example.com -c 'echo stderr >&2; echo stdout'
```

**Reply by FranklinYu** [reply]
I tried the command, and it behaves differently on my machines: both echo shows on terminal, but they are all on local stdout; for example if I do

```sh
ssh -t example.com 'echo stderr >&2; echo stdout' 2>/dev/null
```

I still see both of them. I then did some research, and it seems like [OpenSSH always merge `stdout` and `stderr` when doing `ssh -t`](https://unix.stackexchange.com/questions/134139/stderr-over-ssh-t). I guess I would have to use the aforementioned workaround. Thanks anyway!

(Note: in `ssh`, the `-c` flag means “choosing cipher”, not “command”.)

**Reply by junegunn** [reply]
> (Note: in `ssh`, the `-c` flag means “choosing cipher”, not “command”.)

It was a typo. I missed `sh` while editing the command, so,

```sh
ssh -t example.com sh -c 'echo stderr >&2; echo stdout'
```

**Comment by dimo414** [comment]
Sorry to resurrect an old question @junegunn, but I'm curious if there's a less-simple workaround that doesn't require running `fzf` on the local machine. In my case I'm trying to execute a wrapper script that uses `fzf` remotely, and am hitting the same issue that `ssh -t example.com my_fzf_wrapper` works but `selected_value=$(ssh -t example.com my_fzf_wrapper)` does not.
